pkg/service/controller.go

```diff
@@ -107,19 +149,34 @@ func (c *ControllerService) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}
 
 	// Create DataVolume object
-	dv := &cdiv1.DataVolume{}
-	dv.Name = dvName
-	dv.Namespace = c.infraClusterNamespace
-	dv.Kind = "DataVolume"
-	dv.APIVersion = cdiv1.SchemeGroupVersion.String()
-	dv.ObjectMeta.Labels = c.infraClusterLabels
-	dv.ObjectMeta.Annotations = map[string]string{
-		"cdi.kubevirt.io/storage.deleteAfterCompletion": "false",
-	}
-	dv.Spec.Storage = &cdiv1.StorageSpec{
-		Resources: corev1.ResourceRequirements{
-			Requests: corev1.ResourceList{
-				corev1.ResourceStorage: *resource.NewScaledQuantity(storageSize, 0)},
+	source, err := c.determineDvSource(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	dv := &cdiv1.DataVolume{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "DataVolume",
+			APIVersion: cdiv1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      dvName,
+			Namespace: c.infraClusterNamespace,
+			Labels:    c.infraClusterLabels,
+			Annotations: map[string]string{
+				"cdi.kubevirt.io/storage.deleteAfterCompletion": "false",
+			},
+		},
+		Spec: cdiv1.DataVolumeSpec{
+			Storage: &cdiv1.StorageSpec{
+				Resources: corev1.ResourceRequirements{
+					Requests: corev1.ResourceList{
+						corev1.ResourceStorage: *resource.NewScaledQuantity(storageSize, 0)},
+				},
+				AccessModes: []corev1.PersistentVolumeAccessMode{accessMode},
+				VolumeMode:  volumeMode,
```

Comment:
volume mode should also only be set when we explicitly request RWX

Reply:
The default vslue is nil. If it's not Block, it's not set.

Reply:
rewrote to be more verbose